drivers/adc/adc_shell.c

```diff
@@ -89,32 +48,84 @@ LOG_MODULE_REGISTER(adc_shell);
 #define CMD_HELP_GAIN	"Configure gain.\n"
 #define CMD_HELP_PRINT	"Print current configuration"
 
-#define DEVICES(n) DEVICE_DT_INST_GET(n),
-#define ADC_HDL_LIST_ENTRY(dev_)					\
+#define ADC_HDL_LIST_ENTRY(n)					\
 	{								\
-		.dev = dev_,						\
+		.dev = DEVICE_DT_INST_GET(n),						\
 		.channel_config = {					\
 			.gain = ADC_GAIN_1,				\
 			.reference = ADC_REF_INTERNAL,			\
 			.acquisition_time = ADC_ACQ_TIME_DEFAULT,	\
 			.channel_id = 0,				\
 		},							\
 		.resolution = 0,					\
-	}
-
-#define INIT_MACRO() DT_INST_FOREACH_STATUS_OKAY(DEVICES) NULL
+	},
 
 #define CHOSEN_STR_LEN 20
 static char chosen_reference[CHOSEN_STR_LEN + 1] = "INTERNAL";
 static char chosen_gain[CHOSEN_STR_LEN + 1] = "1";
 
-/* This table size is = ADC devices count + 1 (NA). */
 static struct adc_hdl {
 	const struct device *dev;
 	struct adc_channel_cfg channel_config;
 	uint8_t resolution;
 } adc_list[] = {
-	FOR_EACH(ADC_HDL_LIST_ENTRY, (,), INIT_MACRO())
+#define DT_DRV_COMPAT atmel_sam_afec
+	DT_INST_FOREACH_STATUS_OKAY(ADC_HDL_LIST_ENTRY)
+#undef DT_DRV_COMPAT
```

Comment:
Wouldn't this:
```suggestion
	DT_FOREACH_STATUS_OKAY(atmel_sam_afec, ADC_HDL_LIST_ENTRY)
```
work with `ADC_HDL_LIST_ENTRY` defined as:
```
#define ADC_HDL_LIST_ENTRY(node_id)					\
	{								\
		.dev = DEVICE_DT_GET(node_id),				\
		...
```
?

Reply:
Way better, done.